### PR TITLE
feat(search): search empty states

### PR DESCRIPTION
**Pull request by cyndichin**
## Summary
Adds Search Empty States for both Free and Premium Users

**Free User**
* Saves Online / Offline: Search by Title and URL 
* Archive Online: Search by Title and URL 
* Archive Offline: No Internet Connection
* All Items: Premium Upsell

**Premium User**
* Saves, Archive, All Items Online: Recent Searches
* Saves Offline: Recent Searches
* Archive, All Items Offline: No Internet Connection

Note: Offline will be refined in future tickets, but should appear if you view the search in offline view on Device (simulator is a bit wonky with networking)

## References 
**Free User:** IN-959, IN-990, IN-964
**No Results:** IN-970
**Premium:** IN-966

## Implementation Details
Added more view models that conform to the `EmptyStateViewModel` to handle the empty states for search. Created `SearchViewModel` to handle the logic for our search view and manage the different empty states.

## Test Steps
- [ ] When “Saves” is selected 
AND a free user activates the search experience
AND the device is online
THEN the search results area is empty of results
AND the search results area includes text saying “Search by title or URL”
 
- [ ] When “Archive” is selected 
AND a free user activates the search experience
AND the device is online
THEN the search results area is empty of results
AND the search results area includes text saying “Search by title or URL”
 
- [ ] When “All Items” is selected
AND a free user activates the search experience
AND the device is online
THEN the search results area is empty of results
AND there is a premium upsell visible

- [ ] When “Saves” is selected 
AND a free user activates the search experience
AND the device is offline
THEN the search results area is empty of results
AND the search results area includes text saying “Search by title or URL”

- [ ] When “Archive” is selected 
AND a free user activates the search experience
AND the device is offline
THEN the search results area is empty of results
AND the search results area includes text saying “No Internet Connection"

- [ ] When “All Items” is selected
AND a free user activates the search experience
AND the device is online
THEN the search results area is empty of results
AND there is a premium upsell visible

- [ ] WHEN I submit a search
AND there are no results
AND there has not been an error
THEN display the "No Results" message

- [ ] WHEN I am an online premium user
AND I have no recent searches
THEN the Recent Searches header will appear with a message

- [ ] WHEN I am an offline premium user
THEN I select “All Items” scope
THEN I see a message explaining that I need to have an internet connection to search

## PR Checklist:
- [x] Added Unit / UI tests
- [x] Self Review (review, clean up, documentation, run tests)
- [x] Basic Self QA

## Screenshots
|Premium User|Free User|
|------|-----|
|Premium User|Free User|

### CODE REVIEW

File: Tests iOS/MyList/SearchTests.swift [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        XCTAssertTrue(app.saves.searchView.exists)
+        XCTAssertTrue(app.saves.searchEmptyStateView(for: "search-empty-state").exists)
+
+    }
+
+    func test_search_forPremiumUser_showsEmptyStateView() {
+        tapSearch()
+        XCTAssertTrue(app.saves.searchView.exists)
+        XCTAssertTrue(app.saves.searchEmptyStateView(for: "recent-search-empty-state").exists)
+    }
```
Comment (cyndichin) [inline Tests iOS/MyList/SearchTests.swift:59]: We are testing the different view models via Unit Testing depending on the scope and user type. Do we want to add more cases or those this suffice for now and we can have a more end to end tests as we further implement search?

File: Tests iOS/Support/Elements/SavesElement.swift [hunk 2]
```diff
@@ -48,6 +53,10 @@ struct SavesElement: PocketUIElement {
         element.otherElements[type]
     }
 
+    func searchEmptyStateView(for type: String) -> XCUIElement {
+        element.images[type]
```
Comment (cyndichin) [inline Tests iOS/Support/Elements/SavesElement.swift:57]: this boggles my mind, but the accessibility identifier is found through the image type for the `EmptyStateView`

File: PocketKit/Tests/PocketKitTests/SearchViewModelTests.swift [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    func test_updateScope_forPremiumUser_withOfflineAll_showsSearchEmptyState() {
+        user.status = .premium
+        networkPathMonitor.update(status: .unsatisfied)
+
+        let viewModel = subject()
+        viewModel.updateScope(with: .archive)
+        XCTAssertTrue(viewModel.emptyState is OfflineEmptyState)
+    }
+
```
[the remark this answers is not in the document]
Comment (cyndichin) [reply PocketKit/Tests/PocketKitTests/SearchViewModelTests.swift:99]: thank you for catching this SwiftLint! ❤️

File: PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import Foundation
+import Textile
+
+struct GetPremiumEmptyState: EmptyStateViewModel {
+    let imageAsset: ImageAsset = .diamond
+    let maxWidth: CGFloat = 145
```
Comment (timc-mozilla) [inline PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift:6]: not the biggest fan of hardcoded values. is there a way we can have common dimensions in their own file?

File: PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/RecentSearchEmptyState.swift [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import Foundation
+import Textile
+
+struct RecentSearchEmptyState: EmptyStateViewModel {
+    let imageAsset: ImageAsset = .searchRecent
+    let maxWidth: CGFloat = 145
+    let icon: ImageAsset? = nil
+    let headline = "Recent Searches"
+    let detailText: String? = "Your latest searches will appear here so you can get right back to them"
```
Comment (timc-mozilla) [inline PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/RecentSearchEmptyState.swift:9]: same as above with Chetan's Strings

File: PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/SearchEmptyState.swift [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import Foundation
+import Textile
+
+struct SearchEmptyState: EmptyStateViewModel {
+    let imageAsset: ImageAsset = .search
+    let maxWidth: CGFloat = 145
+    let icon: ImageAsset? = nil
+    let headline = "Search by Title or URL"
```
Comment (timc-mozilla) [inline PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/SearchEmptyState.swift:8]: same as above with Chetan's Strings

File: PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/RecentSearchEmptyState.swift [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import Foundation
+import Textile
+
+struct RecentSearchEmptyState: EmptyStateViewModel {
+    let imageAsset: ImageAsset = .searchRecent
+    let maxWidth: CGFloat = 145
```
Comment (timc-mozilla) [inline PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/RecentSearchEmptyState.swift:6]: same as above with duplicate dimens

File: PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    let icon: ImageAsset? = nil
+    let headline = "Searching in All Your Items?"
+    let detailText: String? = "See results from both saves and archive when you join Pocket Premium."
+    let buttonText: String? = "Get Pocket Premium"
+    let webURL: URL? = URL(string: "https://getpocket.com/premium")
```
[the remark this answers is not in the document]
Comment (cyndichin) [reply PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift:11]: seems out of scope, but lets def make sure to mention this as part of @ckpocket's work!

File: PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift (hunk 4 above)
Comment (cyndichin) [reply PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift:6]: also open to any suggestion on not hardcoding the values, but this was what I could figure out to make the UI look good. Do you mean like a constants file, something like below?
```
struct Constants {
     static let emptyStateMaxWidth: CGFloat  = 145
}
```

Comment (timc-mozilla) [reply PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift:6]: yeah, but if its too much work/not something we do in our codebase, we dont have to introduce it now.

Comment (cyndichin) [reply PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift]: Actually, we do I created one for our margins, so thanks for pointing this out! I updated with the commit `feedback` and will squash when it looks good :) let me know if anything else

File: PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import Foundation
+import Textile
+
+struct GetPremiumEmptyState: EmptyStateViewModel {
+    let imageAsset: ImageAsset = .diamond
+    let maxWidth: CGFloat = 145
+    let icon: ImageAsset? = nil
+    let headline = "Searching in All Your Items?"
+    let detailText: String? = "See results from both saves and archive when you join Pocket Premium."
```
[the remark this answers is not in the document]
Comment (cyndichin) [reply PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift:10]: I only added this to the new classes I created, not the existing classes as it should already be done by the time this gets merged

File: PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (ckpocket) [reply PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift:12]: @timc-mozilla we don't have strings file at this moment, however, once it's added, we will take care of the strings in upcoming work. cc: @cyndichin

File: PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift (hunk 4 above)
Comment (cyndichin) [reply PocketKit/Sources/PocketKit/MyList/EmptyStates/Search/GetPremiumEmptyState.swift]: ugh I made a boo boo, I merged before I squashed, so I squashed this in the feature branch

**Pull request merged**